src/firebase_functions/options.py

```diff
@@ -223,32 +197,141 @@ def asdict_with_global_options(self) -> dict:
         # YAML output as 'null' values.
         return merged_options
 
-
-@_dataclasses.dataclass(frozen=True)
-class _PubSubOptions(_ProviderOptions):
+    def _endpoint(self, **kwargs) -> _manifest.ManifestEndpoint:
+        assert kwargs["func_name"] is not None
+        options_dict = self._asdict_with_global_options()
+        options = self.__class__(**options_dict)
+
+        secret_envs: list[
+            _manifest.SecretEnvironmentVariable] | _util.Sentinel = []
+        if options.secrets is not None:
+            if isinstance(options.secrets, list):
+
+                def convert_secret(
+                        secret) -> _manifest.SecretEnvironmentVariable:
+                    secret_value = secret
+                    if isinstance(secret, SecretParam):
+                        secret_value = secret.name
+                    return {"key": secret_value}
+
+                secret_envs = list(
+                    map(convert_secret, _typing.cast(list, options.secrets)))
+            elif options.secrets is _util.Sentinel:
+                secret_envs = _typing.cast(_util.Sentinel, options.secrets)
+
+        region: list[str] | None = None
+        if isinstance(options.region, list):
+            region = _typing.cast(list, options.region)
+        elif options.region is not None:
+            region = [_typing.cast(str, options.region)]
+
+        vpc: _manifest.VpcSettings | None = None
+        if options.vpc_connector is not None:
+            vpc = ({
+                "connector": options.vpc_connector,
+                "egressSettings": options.vpc_connector_egress_settings.value
+            } if options.vpc_connector_egress_settings is not None else {
+                "connector": options.vpc_connector
+            })
+
+        endpoint = _manifest.ManifestEndpoint(
+            entryPoint=kwargs["func_name"],
+            region=region,
+            availableMemoryMb=options.memory,
+            labels=options.labels,
+            maxInstances=options.max_instances,
+            minInstances=options.min_instances,
+            concurrency=options.concurrency,
+            serviceAccountEmail=options.service_account,
+            timeoutSeconds=options.timeout_sec,
+            cpu=options.cpu,
+            ingressSettings=options.ingress,
+            secretEnvironmentVariables=secret_envs,
+            vpc=vpc,
+        )
+
+        return endpoint
+
+
+@_dataclasses.dataclass(frozen=True, kw_only=True)
+class PubSubOptions(RuntimeOptions):
     """
     Options specific to Pub/Sub function types.
     Internal use only.
     """
 
+    retry: _typing.Optional[bool] = None
+    """
+    Whether failed executions should be delivered again.
+    """
+
     topic: str
     """
     The Pub/Sub topic to watch for message events.
     """
 
-    retry: _typing.Optional[bool] = None
+
+@_dataclasses.dataclass(frozen=True, kw_only=True)
+class DatabaseOptions(RuntimeOptions):
```

Comment:
why `options.DatabaseOptions` instead of `db.Options`? (Again, not sure we can use such a class since it may screw up pydocs)